### ISSUE TITLE
[Bug]: React devtools doesn't attach to devtools on launch, reload needed

### ISSUE THREAD

**Issue by SantiagoChiappe** [electron/electron#41613]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

29.1.4

### What operating system are you using?

Windows

### Operating System Version

Windows 11 Home 23H2

### What arch are you using?

x64

### Expected Behavior

When I launch my Electron app on developer mode, React Devtools are installed and attached to my app.

### Actual Behavior

React Devtools are installed, but not attached to my app. A manual reload or force reload is required to attach it. Same problem if I installed React Devtools manually from Chrome extensions, have to reload manually.
<img width="308" alt="bug01" src="https://github.com/electron/electron/assets/47071792/bbb0c9bd-872f-409a-bcab-6155d94d1ce5">
<img width="681" alt="bug02" src="https://github.com/electron/electron/assets/47071792/a2050880-d0d8-4d42-8488-e600b35c63d5">
![bug03](https://github.com/electron/electron/assets/47071792/30550e3b-306b-4b8f-8516-53951f684b21)
<img width="549" alt="bug04" src="https://github.com/electron/electron/assets/47071792/47826aa8-91b6-4ee3-89b6-accfe807dfa2">

### Additional Information

My code :

```
// Import required modules
const { default: installExtension, REACT_DEVELOPER_TOOLS } = require('electron-devtools-installer');
const path = require('path');
const { app, BrowserWindow } = require('electron');

// Initialize mainWindow variable
let mainWindow;

// Function to create the main window
const createWindow = async () => {
  const isDev = await import('electron-is-dev');

  // Configure the main window
  mainWindow = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      nodeIntegration: true, // Enable Node.js integration
      enableRemoteModule: true, // Enable remote module
      contextIsolation: false, // Disable context isolation
      autoHideMenuBar: true, // Auto-hide menu bar
      devTools: isDev.default, // Show DevTools in development
    },
  });

  // Hide the menu bar
  mainWindow.setMenuBarVisibility(true);

  // Load the appropriate URL based on the environment
  mainWindow.loadURL(
    isDev.default
      ? 'http://localhost:3000' // Development URL
      : `file://${path.join(__dirname, '../build/index.html')}` // Production URL
  );
};

// Create the main window when the app is ready
app.whenReady().then(() => {
  installExtension(REACT_DEVELOPER_TOOLS,  { loadExtensionOptions: { allowFileAccess: true }})
      .then((name) => {
        console.log(`Added Extension: ${name}`);
        // Create the window and the React application
        createWindow();
      })
      .catch((err) => console.log('An error occurred: ', err));
});

// Quit the app when all windows are closed (except on macOS)
app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

// Create a new window when the app is activated (macOS)
app.on('activate', () => {
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow();
  }
  if (mainWindow) {
    mainWindow.reload();
  }
});
```

I tried also reloading the window on my code or adding a timer to reload the window but without success.

**Comment by erikian** [comment]
@SantiagoChiappe are you getting an error similar to this (reported in https://github.com/electron/electron/issues/37876#issuecomment-1984873828) in your main process? 

```
[91044:0308/100700.690484:ERROR:CONSOLE(2)] "Electron sandboxed_renderer.bundle.js script failed to run", source: node:electron/js2c/sandbox_bundle (2)
[91044:0308/100700.690537:ERROR:CONSOLE(2)] "TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))", source: node:electron/js2c/sandbox_bundle (2)
```

**Comment by SantiagoChiappe** [comment]
> @SantiagoChiappe are you getting an error similar to this (reported in [#37876 (comment)](https://github.com/electron/electron/issues/37876#issuecomment-1984873828)) in your main process?
> 
> ```
> [91044:0308/100700.690484:ERROR:CONSOLE(2)] "Electron sandboxed_renderer.bundle.js script failed to run", source: node:electron/js2c/sandbox_bundle (2)
> [91044:0308/100700.690537:ERROR:CONSOLE(2)] "TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))", source: node:electron/js2c/sandbox_bundle (2)
> ```

Yes I do, only when opening devtools. I also got other errors when I reload the app using View -> Reload, but then react devtools appear as I mentioned previously :
```[1] [73616:0410/004950.275:ERROR:CONSOLE(1)] "Request Runtime.evaluate failed. {"code":-32602,"message":"uniqueContextId not found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
[1] [73616:0410/004950.275:ERROR:CONSOLE(1)] "uniqueContextId not found", source: devtools://devtools/bundled/core/sdk/sdk.js (1)
[1] [73616:0410/004950.275:ERROR:CONSOLE(1)] "Extension server error: Inspector protocol error: uniqueContextId not found", source: devtools://devtools/bundled/models/extensions/extensions.js (1)
[1] [73616:0410/004950.278:ERROR:CONSOLE(1)] "Received error with code E_PROTOCOLERROR while checking if react has loaded: "Inspector protocol error: %s"", source: chrome-extension://deachpobkibmlfmamelbofgimdagiiob/build/main.js (1)```

**Comment by louispelarrey** [comment]
Got the same issue

**Comment by Jurin** [comment]
Exact same issue but with Vue devtools with the exception that it will not load even after reloading. Same error when trying to open devtools.

```
[24844:0422/101748.810:ERROR:CONSOLE(1)] "Request Autofill.enable failed. {"code":-32601,"message":"'Autofill.enable' wa
sn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
[24844:0422/101748.903:ERROR:CONSOLE(2)] "Electron renderer.bundle.js script failed to run", source: node:electron/js2c/
renderer_init (2)
[24844:0422/101748.903:ERROR:CONSOLE(2)] "TypeError: object null is not iterable (cannot read property Symbol(Symbol.ite
rator))", source: node:electron/js2c/renderer_init (2)
```

**Comment by khanayaan98** [comment]
Any update on this? how to resolve the issue?

**Comment by alexk111** [comment]
Resolved this by using the `react-devtools` package instead of installing the chrome extension.

1. Exec `yarn add --dev react-devtools`
2. Add `"dev:react-devtools": "react-devtools"` or something like that to the package.json
3. Add `<script src="http://localhost:8097"></script>` to the dev version of your renderer index html file
4. Now when you exec `yarn run dev:react-devtools` it will launch the react-devtools server and make the react devtools available in the app window devtools.

**Comment by lanshenghai** [comment]
> Resolved this by using the `react-devtools` package instead of installing the chrome extension.
> 
> 1. Exec `yarn add --dev react-devtools`
> 2. Add `"dev:react-devtools": "react-devtools"` or something like that to the scripts in the package.json
> 3. Add `<script src="http://localhost:8097"></script>` to the dev version of your renderer index html file
> 4. Now when you exec `yarn run dev:react-devtools` it will launch the react-devtools server and make the react devtools available in the app window devtools.

It's workable on my project. Thanks.

**Comment by nikwen** [comment]
@codebytere You told me you weren't able to reproduce this. Here's an example repo (using the Webpack dev server) where it works reliably for me: https://github.com/nikwen/electron-devtools-bug-repro

Steps to reproduce:

1. `npm start`
2. Open the devtools
3. The React devtools aren't visible
4. Reload the page
5. The React devtools become visible

Here's a video showing the reproduction: https://www.loom.com/share/a1e68289b4594559b01e104c4173f0d0

IIRC this used to work on Electron v28 and broke in Electron v29.

**Comment by codebytere** [comment]
After a bit of suffering and building like 20 different devtools versions, i've determined that https://github.com/facebook/react/commit/f7181993133a89b54f93d79d199be3392e5af6fc broke it on initial load. Not sure quite why yet, but there's a good lead now.

**Comment by Methuselah96** [comment]
I'm working on Redux DevTools to Manifest V3 in https://github.com/reduxjs/redux-devtools/pull/1714, and am seeing similar errors:

```
[28768:0815/225434.455:ERROR:CONSOLE(1)] "Request Autofill.enable failed. {"code":-32601,"message":"'Autofill.enable' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
[28768:0815/225434.455:ERROR:CONSOLE(1)] "Request Autofill.setAddresses failed. {"code":-32601,"message":"'Autofill.setAddresses' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
[28768:0815/225434.559:ERROR:CONSOLE(2)] "Electron sandboxed_renderer.bundle.js script failed to run", source: node:electron/js2c/sandbox_bundle (2)
[28768:0815/225434.559:ERROR:CONSOLE(2)] "TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))", source: node:electron/js2c/sandbox_bundle (2)
[28768:0815/225435.037:ERROR:CONSOLE(2)] "Electron sandboxed_renderer.bundle.js script failed to run", source: node:electron/js2c/sandbox_bundle (2)
```

I'll probably just disable Electron testing for now, since I don't have the bandwidth to support both Manifest V2 and V3. Hoping this will get resolved soon, let me know if I can be of any help!

**Comment by Letrab** [comment]
> After a bit of suffering and building like 20 different devtools versions, i've determined that [facebook/react@f718199](https://github.com/facebook/react/commit/f7181993133a89b54f93d79d199be3392e5af6fc) broke it on initial load. Not sure quite why yet, but there's a good lead now.

It is a very problematic commit for electron, there is a long outstanding issue loading local file urls since that commit as well:

https://github.com/facebook/react/issues/27749

**Comment by nikwen** [comment]
Unfortunately, this is still happening.

**Comment by ftmoose** [comment]
React Devtools not even attaching after refresh on my end, I'm unable to get it working at all. This is with Electron v33.3.0.
I'm on Ubuntu x86.

Here's the code I'm using to load the extension:
```
app.whenReady().then(() => {
  log.info("App Starting")
  log.info(app.getPath("userData"))

  if (isDev) {
    installExtension(REACT_DEVELOPER_TOOLS, { forceDownload: true, loadExtensionOptions: { allowFileAccess: true } })
      .then((ext) => log.debug(`Added Extension: ${ext.name}`))
      .catch((err) => log.debug(`Error adding extension: ${err}`))
  }

  ...
```

Relevant output (same as above):
```
09:19:17.927 › Added Extension: React Developer Tools
[331777:1223/091919.441320:ERROR:CONSOLE(1)] "Request Autofill.enable failed. {"code":-32601,"message":"'Autofill.enable' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
[331777:1223/091919.441356:ERROR:CONSOLE(1)] "Request Autofill.setAddresses failed. {"code":-32601,"message":"'Autofill.setAddresses' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
[331777:1223/091919.585291:ERROR:CONSOLE(2)] "Electron renderer.bundle.js script failed to run", source: node:electron/js2c/renderer_init (2)
[331777:1223/091919.585312:ERROR:CONSOLE(2)] "TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))", source: node:electron/js2c/renderer_init (2)
```

**Comment by Kitenite** [comment]
> React Devtools not even attaching after refresh on my end, I'm unable to get it working at all. This is with Electron v33.3.0. I'm on Ubuntu x86.
> 
> Here's the code I'm using to load the extension:
> 
> ```
> app.whenReady().then(() => {
>   log.info("App Starting")
>   log.info(app.getPath("userData"))
> 
>   if (isDev) {
>     installExtension(REACT_DEVELOPER_TOOLS, { forceDownload: true, loadExtensionOptions: { allowFileAccess: true } })
>       .then((ext) => log.debug(`Added Extension: ${ext.name}`))
>       .catch((err) => log.debug(`Error adding extension: ${err}`))
>   }
> 
>   ...
> ```
> 
> Relevant output (same as above):
> 
> ```
> 09:19:17.927 › Added Extension: React Developer Tools
> [331777:1223/091919.441320:ERROR:CONSOLE(1)] "Request Autofill.enable failed. {"code":-32601,"message":"'Autofill.enable' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
> [331777:1223/091919.441356:ERROR:CONSOLE(1)] "Request Autofill.setAddresses failed. {"code":-32601,"message":"'Autofill.setAddresses' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
> [331777:1223/091919.585291:ERROR:CONSOLE(2)] "Electron renderer.bundle.js script failed to run", source: node:electron/js2c/renderer_init (2)
> [331777:1223/091919.585312:ERROR:CONSOLE(2)] "TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))", source: node:electron/js2c/renderer_init (2)
> ```

Same here

**Comment by nikwen** [comment]
If the React devtools don't appear for you at all on v33, please [file a new issue](https://github.com/electron/electron/issues/new?template=bug_report.yml).

Before doing that, please verify that you're on the most recent version of `electron-devtools-installer` (currently [`v4.0.0`](https://github.com/MarshallOfSound/electron-devtools-installer/releases/tag/v4.0.0)).

**Comment by CatStudioApp** [comment]
still repro on electron v33 with v4 of the installer. creating one here: https://github.com/electron/electron/issues/45128

**Comment by samuelmaddock** [comment]
This issue occurs because [react developer tool's content script](https://github.com/facebook/react/blob/main/packages/react-devtools-extensions/src/contentScripts/prepareInjection.js) attempts to communicate with the background service worker, which isn't running on startup (after the first install).

A fix in Electron will likely be needed to launch these service workers upon startup. For now, you can upgrade to Electron v35 and use the following workaround:

```js
import { app, session as electronSession } from 'electron'

function launchExtensionBackgroundWorkers(session = electronSession.defaultSession) {
  return Promise.all(
    session.getAllExtensions().map((extension) => {
      const manifest = extension.manifest;
      if (manifest.manifest_version === 3 && manifest?.background?.service_worker) {
        return session.serviceWorkers.startWorkerForScope(extension.url);
      }
      return Promise.resolve();
    })
  );
}

app.whenReady().then(async () => {
  await installExtension(REACT_DEVELOPER_TOOLS);
  await launchExtensionBackgroundWorkers();

  // launchMyApp();
});
```

**Comment by samuelmaddock** [comment]
Electron does not call `extensions::ProcessManager::MaybeCreateStartupBackgroundHosts` which is why the service worker neither starts nor invokes the `chrome.runtime.onStartup` listener.

This method can only be called once which works well for Chrome, but might not be best for Electron. I'll think about how this can be addressed within Electron.

**Comment by Stanzilla** [comment]
@nikwen is this enough info or is there anything else that we can provide to get this fixed in electron?

**Comment by nikwen** [comment]
@Stanzilla No extra info needed. As far as I know, @samuelmaddock is working on it.

**Comment by GitMurf** [comment]
> [@Stanzilla](https://github.com/Stanzilla) No extra info needed. As far as I know, [@samuelmaddock](https://github.com/samuelmaddock) is working on it.

@samuelmaddock any updates or is the workaround of essentially loading the extension twice (or similar) still the best workaround today?

Thanks!

**Comment by spamshaker** [comment]
For reference https://github.com/MarshallOfSound/electron-devtools-installer/issues/244#issuecomment-2227173598

**Comment by nikwen** [comment]
@GitMurf This is the best workaround: https://github.com/electron/electron/issues/41613#issuecomment-2644018998

**Comment by nikwen** [comment]
@samuelmaddock Would it make sense to start the service worker as part of `session.extensions.loadExtension()`?

**Comment by samuelmaddock** [comment]
> [@samuelmaddock](https://github.com/samuelmaddock) Would it make sense to start the service worker as part of `session.extensions.loadExtension()`?

@nikwen I think that could be a good approach! Maybe `loadExtension(path, { startBackgroundWorker = true })` so folks can choose whether to opt-out.

**Comment by nikwen** [comment]
Good point! It probably makes sense to make it the default to start the worker, with the ability to opt out.

**Comment by nikwen** [comment]
I've found a better way to fix this: https://github.com/electron/electron/pull/50611

It turns out that the service worker was actually started when an extension was loaded into an app for the very first time.  When the same extension was loaded on subsequent app launches, the service worker would no longer start. The PR makes it so that the service worker is started on subsequent app launches, too.

Because the service worker was already being started on the first app launch, I've decided not to add a configuration option for now.

**Comment by jamilalisgandarov** [comment]
We have the similar issue but for our unpacked extension, where we distribute it ourselves inside our electron wrapper. 

1. We load our unpacked extension
````Electron app loading an unpacked MV3 extension into a per-account Session:
const session = Electron.session.fromPartition(`persist:account-${id}`);
await session.loadExtension(extensionPath); // first install
````

2. Load extension version A: session.loadExtension(pathA) → SW registers, versionId: 4, transitions to running. Works.
3. Bump the extension version on the server, the user clicks "Update" in the app.
4. Our reload path runs:
````
for (const ext of session.getAllExtensions()) {
  session.removeExtension(ext.id);
}
await session.loadExtension(pathB);
````
5. Sometimes, previous versionId: 4 survives, causing extension to work not as expected.

Not all the time, but sometimes it is just not initiating the new service worker for the extension. When checked with service worker listener to get running service workers. I observed that when the issue happens previous service worker session id survives even after remove and reinstall of the extension. We don't have specific workaround for it at the moment, beside closing the browser view and  opening it manually multiple times. 

Electron v36.1.0
Extension - MV3
